### ISSUE TITLE
Rotatable entity selectionboxes

### ISSUE THREAD

**Issue by appgurueu** [minetest/minetest#12372]
## Problem

Many mobs mods such as petz use highly non-square entities (and often fail to take care of properly rotating it). Using a square selectionbox for these both looks bad when highlighted and behaves poorly when pointing at the entities (or when using gun / bow mods that rely on accurate raycasting).

## Solutions

Implement rotatable selectionboxes as an API field `rotate_selectionbox = true or false` (perhaps also as a field `rotate = true or false` in the selectionbox table if the API is to be kept "clean"). If the field is set to `true`, the selectionbox behaves as if it were rotated by the entities rotation. Object properties must be modified for this.

This can be implemented by [inversely rotating the ray around the object](https://github.com/appgurueu/epidermis/blob/be44377d677ae92e89206cfbd5c628ba8d4192fc/paintable.lua#L313-L316). This would have to be implemented in raycasts.

The halo/outline mesh would additionally have to keep track of rotation.

## Alternatives

- [Setting a larger selectionbox that contains the rotated box](https://github.com/appgurueu/epidermis/blob/be44377d677ae92e89206cfbd5c628ba8d4192fc/paintable.lua#L101-L116)
- Setting the smallest selectionbox contained within the rotated box
- Setting a smaller or larger square selectionbox independent of rotation
- Attaching punchable entities to approximate the rotated box using multiple boxes and propagating the punch events

all of which are hardly satisfactory as they can only provide rough approximations

## Additional context

Basically a duplicate of #7652 but creating a new issue was encouraged.

**Comment by ExeVirus** [comment]
### Additional solution
- Viable for _collision and selection boxes_:

**Rotate individual boxes of each selection box table/collision box table, such that they are always Axis Aligned Bounding Boxes (AABB):** 

![Rotations and AABB](https://user-images.githubusercontent.com/44562154/170174395-aae5732e-35e6-4e9f-927e-3c1656630a7f.gif)

**Comment by appgurueu** [comment]
> ### Additional solution
> Rotate individual boxes of each selection box table/collision box table, such that they are always Axis Aligned Bounding Boxes (AABB)

I'm afraid this solution is likely to be inferior both in terms of performance and accuracy for selectionboxes. You don't save any rotations (you still have to rotate the two origins) in comparison to rotating the ray around the entity (which requires rotating the ray direction & origin). Rotations aren't that expensive if matrices are *avoided*, i.e. using quaternions or [Rodrigues' rotation formula](https://en.wikipedia.org/wiki/Rodrigues%27_rotation_formula). **There is ultimately no need to use this approximation for selectionboxes.**

I doubt that it'd work for collisionboxes (irrelevant to this issue) with our current physics code because - as has been pointed out in #10909 already - the code assumes boxes to be static; not to mention that there currently is only single box support (except for attachment hacks).

**Comment by ExeVirus** [comment]
> There is ultimately no need to use this approximation for selectionboxes.

I agree with that assessment, as selection boxes are for raycasts, and the rotation of the ray is far cheaper than the rotation of the boxes. Good point. 

> single box support

I was not aware of this limitation, as I only work with nodes typically and their collision boxes. So for entity collision boxes, you are saying that an entity is only allowed a single box, rather than a table of boxes?

**Comment by appgurueu** [comment]
> I was not aware of this limitation, as I only work with nodes typically and their collision boxes. So for entity collision boxes, you are saying that an entity is only allowed a single box, rather than a table of boxes?

Yes, that's the case.

**Comment by Desour** [comment]
Related: #11679 (About selection meshes, but also requests rotation of those.)

(Some years ago I had the same idea as ExeVirus at #2851 x).)

**Comment by ExeVirus** [comment]
However, I agree with appgurueu on all accounts. 

For this particular use case, selection boxes, as they currently stand with a single box, rotating the raycast is the best route. As well bring oct-trees and multiple boxes and collision boxes into this discussion over time, that answer will change. No need to worry about that now, as the usefulness is already apparent and not too hard to accommodate.
